cpp/src/utilities/type_dispatcher.hpp

```diff
@@ -107,14 +112,83 @@ decltype(auto) type_dispatcher(gdf_dtype dtype,
     case GDF_DATE64:    { return f.template operator()< date64 >(std::forward<Ts>(args)...); }
     case GDF_TIMESTAMP: { return f.template operator()< timestamp >(std::forward<Ts>(args)...); }
     case GDF_CATEGORY:  { return f.template operator()< category >(std::forward<Ts>(args)...); }
-    default:            { assert(false && "type_dispatcher: invalid gdf_type"); }
+    default: { assert(false && "type_dispatcher: invalid gdf_type"); }
   }
   // Need to find out what the return type is in order to have a default return value
   // and solve the compiler warning for lack of a default return
   using return_type = decltype(f.template operator()<int8_t>(std::forward<Ts>(args)...));
   return return_type();
 }
 
-} // namespace cudf
+/**---------------------------------------------------------------------------*
+ * @brief Type-traits struct for mapping a C++ type to it's corresponding
+ * gdf_dtype.
+ *
+ * This traits structure contains a `gdf_dtype` member `value` that returns the
+ * appropriate `gdf_dtype` for the specified C++ type.
+ *
+ * @tparam T The type to map to a `gdf_dtype`
+ *---------------------------------------------------------------------------**/
+template <typename T>
+struct type_to_gdf_dtype {
```

Comment:
Why not a function? 
```
template <typename ColumnElement>
constexpr gdf_dtype gdf_dtype_of() { return GDF_invalid; }
```
we no longer need C++98-style template metaprogramming.

Reply:
I'm just following suit with what the STL uses for type-traits. 

Is the best practice guidance to use `constexpr` functions for type-traits now?

Reply:
Actually, that's a good question to put on StackOverflow. On one hand, remember that most of the standard library was written in the early 1990s where you only had struct-based TMP. And since then we've seen a move towards programming instead of meta-programming as much as possible, not the least of which being:

* `trait_t` instead of `trait::type`
* `trait_v` instead of `trait::value` (C++17)
* `if constexpr` instead of SFINAE+`std::enable_if` (C++17)

But on the other hand, there is the issue of tradition; and the fact that we are _not_ in C++17.

Reply:
Updated to a function.